### PR TITLE
Factor WillPopScope out of Form

### LINKED ISSUE #8293: add onWillPop to Widgets
It would be nice to have onWillPop added to widgets so that it's easier to change default back button behavior. Docs on common usage of this would be helpful as well.

Are you sure you want to Exit? Is one I can think of right now.

### PR DESCRIPTION
**Pull request by abarth**
This functionality is independently useful.

Fixes #8293

### CODE REVIEW

**Comment by abarth** [comment]
@HansMuller

File: packages/flutter/lib/src/widgets/will_pop_scope.dart [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+  @override
+  void dispose() {
+    if (config.onWillPop != null)
+      _route?.removeScopedWillPopCallback(config.onWillPop);
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) => config.child;
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/will_pop_scope.dart]: should we assert that we need a ModalRoute? (in the same way we assert for a Material)

Comment (abarth) [inline packages/flutter/lib/src/widgets/will_pop_scope.dart]: Currently this widget "works" without a modal route in the sense that it does nothing.

**Comment by Hixie** [comment]
test?

LGTM

**Comment by abarth** [comment]
I guess I could test the widget directly, but it should already be tested from Form.

**Pull request merged**